docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+chosen.
+
+###### `share.WriteEDS`
+
+To write EDS into a stream/file `WriteEDS` is introduced.
+// TODO Describe logic
+
+NOTE: CAR provides [a utility](https://github.com/ipld/go-car/blob/master/car.go#L47) to serialize any DAG into the file and
+there is a way to serialize EDS into DAG(`share/ipld.ImportShares`). This approach is the simplest and traverses shares
+and Merkle Proofs in depth-first manner packing them in a CAR file. However, this is incompatible with the requirement
+to truncate the CAR file to get strictly the first quadrant out of it without NMT proofs, so serialization must be
+different from the utility to support that.
+
+NOTE2: Alternatively to `WriteEDS`, and `EDSReader` could be introduced to make EDS-to-stream handling more idiomatic
+and efficient in some cases, with the cost of more complex implementation.
+
+```go
+// WriteEDS writes whole EDS into given io.Writer as CARv1 file.
+// All its shares and recomputed NMT proofs.
+func WriteEDS(context.Context, *rsmt2d.ExtendedDataSquare, io.Writer) error
+```
+
+###### `share.ReadEDS`
+
+To read EDS out of stream/file `ReadEDS` is introduced. Internally, it
+
+- Imports EDS with an empty pre-allocated slice. NOTE: Size can be taken from DataRoot
+- Wraps given io.Reader with [`BlockReader`](https://github.com/ipld/go-car/blob/master/v2/block_reader.go#L16)
+- Reads out blocks one by one and fills up the EDS quadrant via `EDS.SetCell`
+- Recomputes and validates via `EDS.Repair`
+
+```go
+// ReadEDS reads EDS quadrant(1/4) from io.Reader CAR file.
+//
+// It expects strictly first EDS quadrant(top left).
+// Returned EDS is guaranteed to be full and valid against DataRoot, otherwise ReadEDS errors.
+func ReadEDS(context.Context, io.Reader, DataRoot) (*rsmt2d.ExtendedDataSquare, error)
+```
+
+##### `share.EDSStore`
+
+To manage every EDS on the disk FNs/BNs keep `EDSStore` type is introduced in `share` pkg. Each EDS together with its
+Merkle Proofs serializes into CARv1 file. All the serialized CARv1 file blobs are stored as OS FS files as DAGStore
+[Mounts](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go).
+
+The introduced `EDSStore` also maintains (via DAGStore) a top-level index enabling granular and efficient random access
+to every share and/or Merkle proof over every registered CARv1 file. The `EDSStore` provides a custom `Blockstore` interface
+implementation to achieve the access. The main use-case is randomized sampling over the whole chain of EDS block data
+and getting data by namespace.
+
+EDSStore constructor should instantiate
+
+```go
+type EDSStore struct {
+ basepath string 
+ dgstr dagstore.DAGStore
+ topIdx index.Inverted 
+ carIdx index.FullIndexRepo
+ mounts  *mount.Registry
+ ...
+}
+
+// NewEDSStore constructs EDStore over OS directory to store CARv1 files of EDSes and indices for them.
+// Datastore is used to keep inverted/top-level index.
+func NewEDSStore(basepath string, ds datastore.Batching) *EDSStore {
```

Comment:
Can I make this instantiation a little more close to reality or is that out of scope of the ADR? It would only add a few lines and is helpful for understanding I think (adding the FS mount to the registry, passing the relevant things to the EDS store config, and starting the dagstore before returning)

Reply:
oops, not starting the dagstore, since I forgot this will be done by the lifecycle

Reply:
Let's add this. Happy to review your suggestion

Reply:
I extended the constructor a bit, but it still not as full as in your experiment

Reply:
Should be resolved